### PR TITLE
Add inNewFrameWithVars to Protocol.Online

### PR DESCRIPTION
**Pull request by danmatichuk**
This pushes a new assumption frame where the
given bound variables are considered
free variables for the duration of the frame.

### CODE REVIEW

**Comment by danmatichuk** [comment]
This works by adding bindings to fresh free variables that are only live for the current frame. Here we are relying on the cache to enforce our scoping, but I'm not sure if these invented variables can still escape the scope (e.g. by being bound in a cache that's persistent across frames).

**Comment by robdockins** [comment]
This looks right to me, the `DeleteOnPop` mode ought to have the scoping behavior you want.  It would be nice to see some additions to the test suite that exercises this behavior.

File: what4/src/What4/Solver/Yices.hs [hunk 1]
```diff
@@ -436,7 +436,7 @@ newConnection stream ack reqFeatures timeout bindings = do
   let nlSolver = reqFeatures `hasProblemFeature` useNonlinearArithmetic
   let features | efSolver  = useLinearArithmetic
                | nlSolver  = useNonlinearArithmetic .|. useIntegerArithmetic
-               | otherwise = yicesDefaultFeatures
+               | otherwise = reqFeatures
```
Comment (robdockins) [inline what4/src/What4/Solver/Yices.hs]: This looks like the right fix, but we should probably also add `useIntegerArithmetic` to the default features.

Comment (danmatichuk) [reply what4/src/What4/Solver/Yices.hs]: This snuck in by accident, this is orthogonal to the bound variable issue.

**Comment by robdockins** [comment]
Looks good to me

**Pull request merged**